src/crimson/osd/osd.h

```diff
@@ -68,7 +68,7 @@ class OSD : public ceph::net::Dispatcher,
   SimpleLRU<epoch_t, bufferlist, false> map_bl_cache;
   cached_map_t osdmap;
   // TODO: use a wrapper for ObjectStore
-  std::unique_ptr<ceph::os::CyanStore> store;
+  seastar::lw_shared_ptr<ceph::os::CyanStore> store;
```

Comment:
@rzarzynski could you note in the commit message why we want to replace `unique_ptr<>` with `lw_shared_ptr<>`?

Reply:
TODO.
In few words: `PGBackend` needs to use `store` regardless of `OSD` – they stores the pointer on its own. I believe it such situation it's better to use smart-pointer than raw and be responsible for assuring the life-time guarantees manually.

Reply:
@rzarzynski so you think `PGBackend` could outlive `OSD`? but i think OSD owns PG, and `PG` should own `PGBackend`, am i right?

Reply:
> i think OSD owns PG, and PG should own PGBackend, am i right?

That's also my thinking about the relationships we have in our minds (the *intention*). However, when I decouple from it to focus only on the implementation, I can see:
* `OSD::pgs` using `Ref<PG>` which suggest IMHO a relationship weaker than ownership,
* unwritten contract between `PGBackend` and `OSD`. The former draws assumption that `store` will be accessible over its entire life. However, raw pointer doesn't say much, so the obligation must be both discovered and verified basing solely on human. 

Actually it's really unlikely that something like `PG` outlives `OSD`. I don't have any reason to think the current implementation is buggy. The patches are nothing more than a clean-up.

Reply:
it's just confusing from my perspective. as my impression is that the ObjectStore is a critical property of its OSD. by claiming it as a shared_ptr, we are telling the reader that someone else is *holding* it also. actually i don't think

> it stores the pointer on its own

is true. `PGBackend` is using the service offered by OSD. the reason it has a pointer is just because it would be easier for it to dereference the object store. as an alternative approach, we could let `PGBackend` hold a pointer of `OSD`, and add a method like `OSD::get_store()`. because OSD not only offers the access to osdmap, it also offers the access to the underlying store.

FWIW, monitor is also using this design -- each `PaxosService` holds a pointer to `Monitor`. and `Monitor` holds a `vector<std::unique_ptr<PaxosService>>`.

the reason why i don't like it is that, i just don't feel right when a member variable keeps a pointer to its owner -- this kind of coupling renders it more difficult to write a unit test.

so if we really want this kind of ownership more explicit, probably we could introduce a class named StorageProvider? and let `OSD` to implement it?

Reply:
There's a bigger reason to assume that OSD > PG > PGBackend.  osd::start and osd::stop mount() and umount() store respectively.  Even if a PGBackend object and a store object manage to outlive OSD, it still wouldn't be valid for PGBackend to access the store pointer.